### ISSUE TITLE
emacs render delay without X

### ISSUE THREAD

**Issue by ebayer** [sedwards2009/extraterm#285]
Hello

When running emacs with no X inside extraterm, emacs always waits for nearly 2 seconds to show up, no matter what. This does not happen with any other terminal programs. I also tried disabling bash integration and framing output, or enabling "reduce graphic effects" under general settings.

Here are the numbers, showing time to start the application on same machine, without loading any startup options at all.

Running emacs in KDE Konsole:

```
time /usr/bin/emacs -nw -Q -e kill-emacs

real    0m0,075s
user    0m0,024s
sys     0m0,004s
```

Running emacs in Extraterm:

```
time /usr/bin/emacs -nw -Q -e kill-emacs

real    0m2,362s
user    0m0,015s
sys     0m0,012s
```

Also I realized this happens if I do not use X Window, when using X window, render times are much better, although not as good as running in another terminal.

```
time -p /usr/bin/emacs -Q -e "kill-emacs"
real 1.14
user 0.93
sys 0.08

time -p /usr/bin/emacs -nw -Q -e "kill-emacs"
real 2.47
user 0.43
sys 0.09
```

I do not know how to debug this further. Let me know if I can try anything else.

**Comment by sedwards2009** [comment]
What exactly do you mean by "no X"? no `DISPLAY` variable? or Wayland?

**Comment by ebayer** [comment]
Sorry, I meant this:

https://centos.pkgs.org/7/centos-x86_64/emacs-nox-24.3-23.el7.x86_64.rpm.html

Emacs, running without any X window support, clearly shows this problem. When started with X Window support from inside Extraterm, render times are a little bit better, but way more worse than running in any other terminal emulator.

Would running emacs with strace and attaching output help you debug this?

**Comment by sedwards2009** [comment]
What I see here is that when emacs starts up it displays the top and bottom bars, then waits a couple seconds, and then shows the welcome screen. The other terminal I have here does not act the same. Does this match what you see?

I'm not trying to figure out what the difference is. It could be that emacs is waiting for some response from the terminal and timing out instead.

**Comment by sedwards2009** [comment]
I just noticed that if you start emacs and then quick hit another key, then it starts up faster. emacs is waiting for some kind of automatic response from the terminal.

**Comment by ebayer** [comment]
Yes, this is the same behaviour I see. Can confirm that when I hit a key, timeout cancels.

Here is the timed out syscall:

```
strace -o strace.out -ttTC /usr/bin/emacs -Q -e "kill-emacs"
...
10:52:38.706945 rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0 <0.000011>
10:52:38.706980 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0 <0.000014>
10:52:38.707017 pselect6(7, [6], [], NULL, {tv_sec=1, tv_nsec=999893766}, {NULL, 8}) = 0 (Timeout) <2.002082>
10:52:40.709265 rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0 <0.000120>
10:52:40.709482 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0 <0.000013>                                                                                                                                    
10:52:40.709533 ioctl(6, FIONREAD, [0]) = 0 <0.000015>
```

That pselect6 call waits for 2 secons. Here is the same strace lines with another terminal:

```
10:57:23.123340 rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0 <0.000008>
10:57:23.123368 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0 <0.000008>
10:57:23.123393 pselect6(6, [5], [], NULL, {tv_sec=1, tv_nsec=996049424}, {NULL, 8}) = 1 (in [5], left {tv_sec=1, tv_nsec=954858531}) <0.041248>                                                           
10:57:23.164757 --- SIGIO {si_signo=SIGIO, si_code=SI_KERNEL} ---
10:57:23.164854 rt_sigreturn({mask=[]}) = 1 <0.000129>
10:57:23.165085 rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0 <0.000013>
10:57:23.165135 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0 <0.000012>
10:57:23.165174 ioctl(5, FIONREAD, [11]) = 0 <0.000013>
```

**Comment by sedwards2009** [comment]
I was hunting this down last night and got much closer to a cause.

**Comment by sedwards2009** [comment]
Fix released in 0.54.0
